src/client/action/list_databases.rs

```diff
@@ -0,0 +1,140 @@
+use std::{future::IntoFuture, marker::PhantomData};
+
+use bson::{Document, Bson};
+use futures_util::FutureExt;
+
+use crate::{Client, ClientSession, operation::list_databases as op, error::{Result, ErrorKind}, results::DatabaseSpecification, client::BoxFuture};
+
+impl Client {
+    /// Gets information about each database present in the cluster the Client is connected to.
+    pub fn list_databases(
+        &self,
+    ) -> ListDatabases {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+
+    /// Gets the names of the databases present in the cluster the Client is connected to.
+    pub fn list_database_names(
+        &self,
+    ) -> ListDatabases<'_, Names> {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+}
+
+/// Gets information about each database present in the cluster the Client is connected to.
+#[must_use]
+pub struct ListDatabases<'a, M: Mode = Full> {
+    client: &'a Client,
+    options: Option<op::Options>,
+    session: Option<&'a mut ClientSession>,
+    mode: PhantomData<M>,
+}
+
+
+mod private {
+    pub trait Sealed {}
+    impl Sealed for super::Full {}
+    impl Sealed for super::Names {}    
+}
+
+pub trait Mode: private::Sealed {}
+impl Mode for Full {}
+impl Mode for Names {}
+
+pub struct Full;
+pub struct Names;
+
+impl<'a, M: Mode> ListDatabases<'a, M> {
+    fn options(&mut self) -> &mut op::Options {
+        self.options.get_or_insert_with(op::Options::default)
+    }
+
+    #[cfg(test)]
+    pub(crate) fn with_options(mut self, value: impl Into<Option<op::Options>>) -> Self {
```

Comment:
minor concern: we won't actually be testing the builder codepath if we're setting options via this method in tests. probably not a huge deal since the setters are very straightforward, but it is worth considering whether we want any test coverage of those methods

Reply:
Yeah, good to think about, I don't think it's worth testing specifically though.  I actually suspect that once I've written a couple of these I'll be able to put together a declarative macro to automate the boilerplate, which would make tests even less useful.